File: Daily/2023/1802_maximumValueAtAGivenIndexInABoundedArray.py
from typing import *


class Solution:
    def getSum(self, height, n, index):
        # mid
        ttl = height

        # to left
        if height - index > 0:
            ttl += ((height - index) + (height - 1)) * index // 2
        else:
            ttl += index - height + 1
            ttl += (1 + height-1) * (height - 1) // 2

        # to right
        if height - (n - 1 - (index+1) + 1) > 0:
            ttl += ((height - (n - 1 - (index+1) + 1)) + (height - 1)) * (n - 1 - (index+1) + 1) // 2
        else:
            ttl += (n - 1 - (index+1) + 1) - height + 1
            ttl += (1 + height - 1) * (height - 1) // 2

        return ttl

    def maxValue(self, n: int, index: int, maxSum: int) -> int:
        left, right = 1, maxSum
        while left < right:
            mid = (left + right) // 2 + 1

            if self.getSum(mid, n, index) > maxSum:
                right = mid - 1
            else:
                left = mid
        
        return left

File: Daily/2023/1802_maximumValueAtAGivenIndexInABoundedArray.cs
public class Solution {
    private long GetSum(long height, int n, int index) {
        long ttl = 0;
        if (height > index) {
            ttl += (height - index + height) * (index + 1) / 2;
        }
        else {
            ttl += index - height + 1;
            ttl += (1 + height) * height / 2;
        }

        if (height > n - index) {
            ttl += (height - (n - index) + 1 + height) * (n - index) / 2;
        }
        else {
            ttl += n - (index + height);
            ttl += (1 + height) * height / 2;
        }

        return ttl - height;
    }

    public int MaxValue(int n, int index, int maxSum) {
        int left = 1, right = maxSum;
        while (left < right) {
            int mid = right - (right - left) / 2;

            if (GetSum(mid, n, index) <= maxSum) {
                left = mid;
            }
            else {
                right = mid - 1;
            }
        }
        
        return left;
    }
}